fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java

```diff
@@ -503,6 +512,59 @@ protected void setOptimizeForIndexFilters(boolean shouldOptimizeForIndexFilters)
                 .build());
     }
 
+    /**
+     * For the cascades planner, plan the query, serialize the plan to bytes, parse those bytes and reconstruct the plan.
+     * @param query the query
+     * @return the plan that is either just planned or planned, then serialized, deserialized, and reconstructed
+     */
+    @Nonnull
+    protected RecordQueryPlan planQuery(@Nonnull final RecordQuery query) {
+        return planQuery(this.planner, query);
+    }
+
+    /**
+     * For the cascades planner, plan the query, serialize the plan to bytes, parse those bytes and reconstruct the plan.
+     * @param planner the planner to use
+     * @param query the query
+     * @return the plan that is either just planned or planned, then serialized, deserialized, and reconstructed
+     */
+    @Nonnull
+    protected static RecordQueryPlan planQuery(@Nonnull final QueryPlanner planner, @Nonnull final RecordQuery query) {
+        final RecordQueryPlan plannedPlan = planner.plan(query);
+        if (planner instanceof RecordQueryPlanner) {
+            return plannedPlan;
+        }
+        Assertions.assertTrue(planner instanceof CascadesPlanner);
+        return verifySerialization(plannedPlan);
+    }
+
+    /**
+     * Serialize the plan to bytes, parse those bytes, reconstruct, and compare the deserialized plan against the
+     * original plan.
+     * @param plan the original plan
+     * @return the deserialized and verified plan
+     */
+    @Nonnull
+    protected static RecordQueryPlan verifySerialization(@Nonnull final RecordQueryPlan plan) {
+        PlanSerializationContext serializationContext = new PlanSerializationContext(new DefaultPlanSerializationRegistry(),
+                PlanHashable.CURRENT_FOR_CONTINUATION);
+        final RecordQueryPlanProto.PRecordQueryPlan planProto = plan.toRecordQueryPlanProto(serializationContext);
+        final byte[] serializedPlan = planProto.toByteArray();
+        final RecordQueryPlanProto.PRecordQueryPlan parsedPlanProto;
+        try {
+            parsedPlanProto = RecordQueryPlanProto.PRecordQueryPlan.parseFrom(serializedPlan);
+        } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException(e);
+        }
+
+        serializationContext = new PlanSerializationContext(new DefaultPlanSerializationRegistry(), PlanHashable.CURRENT_FOR_CONTINUATION);
+        final RecordQueryPlan deserializedPlan =
+                RecordQueryPlan.fromRecordQueryPlanProto(serializationContext, parsedPlanProto);
+        Assertions.assertEquals(plan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION), deserializedPlan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION));
+        Assertions.assertTrue(plan.structuralEquals(deserializedPlan));
+        return deserializedPlan;
```

Comment:
One interesting quirk here is that this means that we are always running the deserialized plans from Cascades. Theoretically, this means that we could face a bug where, for instance, a plan is broken unless we first serialize and deserialize it (e.g., there's some state that gets set up correctly during plan deserialization but is in an uninitialized state before). I don't really have a _great_ solution for it, though. We could randomly return the serialized or deserialized plan, though that wouldn't be repeatable. We could modify `executeQuery` so that it executes the plan both before serialization and after serialization and validates the same results get back, or something, though that's also a bit iffy.

This shouldn't be a concern for now, because the tests were already validating that the non-serialized plan works, but it could be a problem in the future